### PR TITLE
[IntroCSPy-week5] - Contains exercises for Week 5

### PR DESCRIPTION
**Pull request by kayleabox**
This PR has work for week 5 exercises. There is still some cleaning up to been done and tests that should probably be added. 

ps6 directory was originally provided by the class. I left the layout of files in this directory as they were provided and simply filled out the requested functions. The cipher directory contains my attempt to do a little more reorganizing of the classes and functions provided by the class. I only added tests for files in cipher directory.
 
To run files in ps6:
``python3 week5/solutions/ps6/ps6.py`` 

To run cipher project: 
``python3 week5/solutions/cipher/run_cipher.py``

To run cipher tests:
`` python3 week5/solutions/cipher/run_tests.py``

To run files in short_problems:
``python3 week5/solutions/short_problems/generator.py``

### CODE REVIEW

File: week5/solutions/cipher/Message.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        + list(self.map_alphabet(alphabet.upper(), shift).items()))
+
+    def map_alphabet(self, alphabet, shift):
+        return {letter: alphabet[(alphabet.index(letter) + shift) - 26] for letter in alphabet}
+
+    def apply_shift(self, shift):
+        if 0 <= shift < 26:
+            cipher_dict = self.build_shift_dict(shift)
+            return ''.join([cipher_dict[char] if char.isalpha() else char for char in self.get_message_text()])
+        #return ''.join([self.build_shift_dict(shift)[char] if char.isalpha() else char for char in self.get_message_text() if 0 <= shift < 26])
```
Comment (kayleabox) [inline week5/solutions/cipher/Message.py]: I am not sure which option is better here. I suppose that setting self.build_shift_dict(shift) to a variable that has a descriptive name makes this easier to read and makes the return statement a little shorter/easier to read. But it is possible to make the function a little shorted by not setting self.build_shift_dict(shift) to a variable.

File: week5/solutions/cipher/Message.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    def get_message_text(self):
+        return self.message_text
+
+    def get_valid_words(self):
+        return self.valid_words[:]
+        
+    def build_shift_dict(self, shift):
+        alphabet = string.ascii_lowercase
+        return dict(list(self.map_alphabet(alphabet, shift).items())
```
Comment (kayleabox) [inline week5/solutions/cipher/Message.py]: Not sure if it is worth declaring an alphabet variable or better to just pass string.ascii_lowercase and string.ascii_uppercase directly to the calls to map_alphabet.
```
def build_shift_dict(self, shift):
  return dict(list(self.map_alphabet(string.ascii_lowercase, shift).items()) 
  + list(self.map_alphabet(string.ascii_uppercase, shift).items()))
```

**Comment by inchoate** [comment]
@kayleabox You will need to fix your `.gitignore` as it's conflicting and won't admit a merge.

File: week5/solutions/cipher/Message.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import string
```
Comment (inchoate) [inline week5/solutions/cipher/Message.py]: The `module` is typically called `message.py` and the class `Message`. Thus you'd end up writing:
```
from message import Message
```

File: week5/solutions/cipher/CipherTextMessage.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    def get_decrypted_text_array(self):
+        return [self.apply_shift(n) for n in range(0, 25)]
+
+    def create_real_words_dict(self, decrypted_text):
+        return {decrypted_text.index(element) : self.count_real_words(element.split(' ')) for element in decrypted_text}
+
+    def count_real_words(self, new_words):
+        real_words = 0 
+        return [real_words + 1 for word in new_words if self.is_word(self.get_valid_words(), word)]
```
Comment (inchoate) [inline week5/solutions/cipher/CipherTextMessage.py]: Thus function sounds like it returns an `int` b/c it's a `count` something function. You return a list. Why not:
```
return len([word in new_words if self.is_word(self.get_valid_words(), word)]
```

Maybe I'm missing something.

This code, while, I think correct, is a little un-Pythonic in that it's hard to read.

File: week5/solutions/cipher/Message.py (hunk 2 above)
Comment (inchoate) [reply week5/solutions/cipher/Message.py]: The name `string.ascii_lowercase` is clear enough and short enough. I'd just leave that one and avoid the new line of code.

File: week5/solutions/cipher/Message.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    def get_message_text(self):
+        return self.message_text
+
+    def get_valid_words(self):
+        return self.valid_words[:]
+        
+    def build_shift_dict(self, shift):
+        alphabet = string.ascii_lowercase
+        return dict(list(self.map_alphabet(alphabet, shift).items()) 
+        + list(self.map_alphabet(alphabet.upper(), shift).items()))
```
Comment (inchoate) [inline week5/solutions/cipher/Message.py]: You have two list creations that use almost the exact same code. Can you simplify?

File: week5/solutions/cipher/Message.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    def map_alphabet(self, alphabet, shift):
+        return {letter: alphabet[(alphabet.index(letter) + shift) - 26] for letter in alphabet}
```
Comment (inchoate) [inline week5/solutions/cipher/Message.py]: Here's how I like to do these:
```
def map_alphabet(self, alphabet, shift):
    def mapped_letter(letter):
        return alphabet[(alphabet.index(letter) + shift) - 26]
    return {letter: mapped_letter(letter) for letter in alphabet}
```

You create an internal, named function. This way the logic is in the function, and the assignment is clean and more semantic.

File: week5/solutions/cipher/Message.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    def apply_shift(self, shift):
+        if 0 <= shift < 26:
+            cipher_dict = self.build_shift_dict(shift)
+            return ''.join([cipher_dict[char] if char.isalpha() else char for char in self.get_message_text()])
```
Comment (inchoate) [inline week5/solutions/cipher/Message.py]: Sometimes making these multi-line helps:
```
return ''.join([cipher_dict[char]                                 # assignment logic here
                       if char.isalpha() else char                   # filter logic here
                       for char in self.get_message_text() . # iterator logic here;
                     ])
```

Not perfect, but easier to read sometimes.

File: week5/solutions/cipher/load_words.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import os
+
+current_dir = os.path.dirname(os.path.abspath(__file__))
+WORDLIST_FILENAME = os.path.join(current_dir, 'words.txt')
+
+def load_words():
+    with open(WORDLIST_FILENAME, 'r') as in_file:
```
Comment (inchoate) [inline week5/solutions/cipher/load_words.py:7]: Nice!

File: week5/solutions/short_problems/generator.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+def genPrimes():
```
Comment (inchoate) [inline week5/solutions/short_problems/generator.py:1]: Nice -- generators!

**Review by inchoate — COMMENTED** [review]
A few small comments.

Nice work, @kayleabox .

File: week5/solutions/cipher/CipherTextMessage.py (hunk 4 above)
Comment (kayleabox) [reply week5/solutions/cipher/CipherTextMessage.py]: I don't think you are missing anything. :) Thanks for pointing this out! I ended up with this:
```
    def count_real_words(self, new_words):
        return len([word for word in new_words if self.is_word(self.get_valid_words(), word)])
```

File: week5/solutions/cipher/Message.py (hunk 6 above)
Comment (kayleabox) [reply week5/solutions/cipher/Message.py]: So, I want to be able to do something like: 
```
 def map_alphabet(self, alphabet, shift):
     def mapped_letter(letter):
         return alphabet[(alphabet.index(letter) + shift) - 26]
     return {letter: mapped_letter(letter), letter.upper(): mapped_letter(letter).upper() for letter in alphabet}
```
in order to make the dictionary in one call to the map_alphabet function. Then I could make build_shift_dict less redundant by updating it to: 
```
    def build_shift_dict(self, shift):
        return self.map_alphabet(string.ascii_lowercase, shift)
```
But, from what I have read, it appears that I cannot return two key-value pairs at one time in a dictionary comprehension. Is there a way for me to do something like that without nesting for loops? 

I found this solution with nested for loops. I am not sure if it is best to pass in the array or create it in the for loop itself.
```
    def build_shift_dict(self, shift):
        return self.map_alphabet([string.ascii_lowercase, string.ascii_uppercase], shift)

    def map_alphabet(self, alphabets, shift):
        def mapped_letter(letter, alphabet):
          return alphabet[(alphabet.index(letter) + shift) - 26]
        return {letter: mapped_letter(letter, alphabet) for alphabet in alphabets for letter in alphabet}
```

This also worked, but has a few extra variables:
```   
 def map_alphabet(self, alphabet, shift):
        def mapped_letter(letter):
            return alphabet[(alphabet.index(letter) + shift) - 26]

        mapped_alphabet = {} 
        for letter in alphabet:
            mapped_alphabet[letter] = mapped_letter(letter)
            mapped_alphabet[letter.upper()] = mapped_letter(letter).upper()
        return mapped_alphabet
```

Do any of these seem like decent solutions? Or were you thinking something totally different in terms of me simplifying build_shift_dict that was unrelated to the map_alphabet function?

File: week5/solutions/cipher/cipher_text_message.py [hunk 10]
```diff
@@ -1,7 +1,7 @@
 import os
 import string
 
-from Message import Message
+from message import Message
```
Comment (inchoate) [inline week5/solutions/cipher/cipher_text_message.py]: Did you `git mv Message.py message.py`?

Comment (kayleabox) [reply week5/solutions/cipher/cipher_text_message.py]: I thought I did. But it appears I did not commit the change.

**Review by inchoate — APPROVED** [review]

**Pull request merged**